pom.xml

```diff
@@ -381,6 +381,66 @@
         <scala.compat.version>2.13</scala.compat.version>
       </properties>
     </profile>
+
+    <profile>
+      <id>java21</id>
+      <activation>
+        <jdk>[21,)</jdk>
+      </activation>
+
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <executions>
+              <execution>
+                <id>java21-multi-version-compile</id>
+                <phase>compile</phase>
+                <goals>
+                  <goal>compile</goal>
+                </goals>
+                <configuration>
+                  <release>21</release>
+                  <compileSourceRoots>
+                    <compileSourceRoot>src/main/java21</compileSourceRoot>
+                  </compileSourceRoots>
+                  <multiReleaseOutput>true</multiReleaseOutput>
+                </configuration>
+              </execution>
+              <execution>
+                <id>java21-multi-version-services</id>
+                <phase>compile</phase>
+                <goals>
+                  <goal>compile</goal>
+                </goals>
+                <configuration>
+                  <release>21</release>
+                  <compileSourceRoots>
+                    <compileSourceRoot>src/main/java21</compileSourceRoot>
+                  </compileSourceRoots>
+                  <proc>only</proc>
+                  <annotationProcessorPaths>
+                    <path>
+                      <groupId>org.glassfish.hk2</groupId>
+                      <artifactId>hk2-metadata-generator</artifactId>
+                    </path>
+                    <path>
+                      <groupId>org.apache.logging.log4j</groupId>
+                      <artifactId>log4j-core</artifactId>
+                    </path>
+                    <path>
+                      <groupId>com.google.auto.service</groupId>
+                      <artifactId>auto-service</artifactId>
+                    </path>
+                  </annotationProcessorPaths>
```

Comment:
This is a change we should add in Pinot even if this is never merged

Reply:
Added in #14185